### PR TITLE
Fixes #162: Connects AudioPlayerController to AudioFragment

### LINKED ISSUE #162: UI Structure: AudioFragment
This is tracking implementing the UI structure for the audio bar fragment.

### PR DESCRIPTION
**Pull request by jamesxu0**
## Explanation

Fixes #162: Connects AudioPlayerController to AudioFragment by observing live data that AudioPlayerController sends.

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is **assigned** to an appropriate reviewer.

### CODE REVIEW

**Comment by jamesxu0** [comment]
I'm realized that ViewModels don't get recreated onConfigurationChanges, so the UI automatically retains it position and current play state through screen rotations. In onDestroy, I only release the MediaPlayer if it is not a configuration change. This allows constant audio to be playing for screen rotations.

**Comment by jamesxu0** [comment]
I couldn't import robolectric into my AudioFragment test case, which important because I need a reference to ShadowMediaPlayer to invoke certain callbacks for the UI to actually be interactable. I ran into problems with duplicated classes in "org.codehaus" after following this SO post https://stackoverflow.com/questions/57594179/how-to-resolve-duplicated-gradle-dependency-issues/57819278#57819278. Actually, I not sure how espresso interacts with the MediaPlayer setting the data source, but I'm assuming it doesn't actually make network calls.

**Comment by jamesxu0** [comment]
I couldn't import robolectric into my AudioFragment test case, which important because I need a reference to ShadowMediaPlayer to invoke certain callbacks for the UI to actually be interactable. I ran into problems with duplicated classes in "org.codehaus" after following this SO post https://stackoverflow.com/questions/57594179/how-to-resolve-duplicated-gradle-dependency-issues/57819278#57819278. Also, I not sure how espresso interacts with the MediaPlayer setting the data source, but I assume it doesn't actually make network calls, so the prepared callback will never fire.

File: domain/src/main/assets/welcome.json
Comment (BenHenning) [inline domain/src/main/assets/welcome.json]: I think we need to point to Oppia's GCS bucket since normally file names are just the flenames, not fully qualified. Otherwise the real implementation downloading the stream won't work correctly.

Comment (jamesxu0) [reply domain/src/main/assets/welcome.json]: Yeah I know. I was just putting full http links for testing purposes. I'm still waiting on Sean's response on how to get the GCS_RESOURCE_BUCKET_NAME.

File: app/src/main/java/org/oppia/app/player/audio/AudioFragment.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/audio/AudioFragment.kt]: Here & elsewhere: prefer to use ``requireActivity()`` since we only support injectable fragments that have activities (hence why we override onAttach).

File: app/src/main/java/org/oppia/app/player/audio/AudioFragment.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/audio/AudioFragment.kt]: We should do this check in the presenter, instead. I'd expect handleOnDestroy to always be called according to the onDestroy guidelines, not some subset of that lifecycle. Bugs could arise from incorrect assumptions in the presenter about when onDestroy is called.

Instead, we should gate whatever logic shouldn't run when a configuration change is triggering destroy using this check, instead.

Comment (BenHenning) [reply app/src/main/java/org/oppia/app/player/audio/AudioFragment.kt]: Ditto for other lifecycle methods.

File: app/src/main/java/org/oppia/app/player/audio/AudioFragment.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/audio/AudioFragment.kt]: Suggest either using a single line comment, or duplicating to two KDoc comments (doc comments are always associated to the element they are next to, so as it stands ``getUserIsSeeking`` is being documented with this comment, which might be a bit confusing in generated documentation).

File: app/src/main/java/org/oppia/app/player/audio/AudioFragmentPresenter.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/audio/AudioFragmentPresenter.kt:53]: Nit: here & below, please separate methods with a single blank newline per https://google.github.io/styleguide/javaguide.html#s4.6.1-vertical-whitespace.

File: app/src/main/java/org/oppia/app/player/audio/AudioFragmentPresenter.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/audio/AudioFragmentPresenter.kt]: This isn't really a getter since it's not returning anything (and the KDoc confusingly suggests it's a setter). Consider instead: ``retrieveVoiceoverMappings`` & update the KDoc accordingly.

File: app/src/main/java/org/oppia/app/player/audio/AudioFragmentPresenter.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/audio/AudioFragmentPresenter.kt]: Prefer passing the error to the log method to ensure the stack trace is also logged, e.g.:

```kotlin
logger.e("AudioFragment", "Failed to retrieve Exploration: $explorationId", explorationResult.getErrorOrNull())
```

File: app/src/main/java/org/oppia/app/player/audio/AudioFragmentPresenter.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/audio/AudioFragmentPresenter.kt]: Nit: Prefer wrapping the ``if`` if it has side effects like this, e.g.:

```kotlin
if (fromUser) {
  userProgress = progress
}
```

File: app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt]: If this case is illegal, should it throw an exception?

File: app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt]: I wonder if we can make the name more clear so that it's clearer at callsites what calling this method actually does. Though when reading this, I'm wondering if we actually want to call it in the places that we are. Perhaps we really want two methods: one that handles toggling play/pause in the UI, and another that properly auto starts/stops it based on lifecycles. That will result in more predictability, I think.

File: app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt]: Ditto for this ``if`` (prefer to wrap here & below).

File: app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt]: Can you send an email to Ankit to ask about this? Let's try to get this unblocked ASAP since we should resolve this before checking in the code (otherwise it won't work). @oppia/dev-workflow-team & @oppia/core-maintainers in case anyone knows where this value comes from (I couldn't find it in oppia/oppia).

File: domain/src/main/assets/welcome.json
Comment (BenHenning) [reply domain/src/main/assets/welcome.json]: See my comment above--let's see if we can resolve this before Sean returns.

File: domain/src/main/java/org/oppia/domain/audio/AudioPlayerController.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/audio/AudioPlayerController.kt:93]: Why is recreation necessary?

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationDataController.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/exploration/ExplorationDataController.kt:1]: Please update your branch to the latest develop--I think these changes need to move to ``ExplorationRetriever``.

File: domain/src/test/java/org/oppia/domain/audio/AudioPlayerControllerTest.kt
Comment (BenHenning) [inline domain/src/test/java/org/oppia/domain/audio/AudioPlayerControllerTest.kt:262]: Don't we still want to verify this?

File: app/src/sharedTest/java/org/oppia/app/player/audio/AudioFragmentTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/player/audio/AudioFragmentTest.kt]: Is this still blocking you? Please resolve before finishing this PR.

File: app/src/sharedTest/java/org/oppia/app/player/audio/AudioFragmentTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/player/audio/AudioFragmentTest.kt:55]: Please add remaining tests.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @jamesxu0! Mostly nits, but my main concerns were us not properly pulling the audio files from GCS, and not thoroughly testing the UI changes.

File: domain/src/main/java/org/oppia/domain/audio/AudioPlayerController.kt
Comment (jamesxu0) [reply domain/src/main/java/org/oppia/domain/audio/AudioPlayerController.kt:93]: This recreation is necessary because you can't reset the mediaplayer after being released.

File: domain/src/test/java/org/oppia/domain/audio/AudioPlayerControllerTest.kt
Comment (jamesxu0) [reply domain/src/test/java/org/oppia/domain/audio/AudioPlayerControllerTest.kt:262]: Updated the test to check that the old live data does not get new updates

**Comment by jamesxu0** [comment]
@BenHenning PTAL. Addressed all comments and added more test cases.

File: utility/src/main/java/org/oppia/util/gcsresource/GcsResourceModule.kt
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/gcsresource/GcsResourceModule.kt]: Nit: please add newline at the end of the file.

File: utility/src/main/java/org/oppia/util/gcsresource/GcsResouceAnnotations.kt
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/gcsresource/GcsResouceAnnotations.kt:1]: Is this & the next file reversed? It seems like this contains the module despite being a file called annotations, and the module file contains an annotation. Please fix.

File: app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt]: Nit: please remove extra newline.

File: app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt]: Consider UiAudioPlayStatus instead of just AudioPlayStatus. That might help further differentiate from PlayStatus.

File: app/src/main/res/values/strings.xml
Comment (BenHenning) [inline app/src/main/res/values/strings.xml]: This is what the screenreader will read when the user has the button selected. It might be nicer to emphasize the action first, e.g.

"Play audio" / "Pause audio"

I also think that in context 'button' is probably redundant and thus could be omitted.

File: app/src/sharedTest/java/org/oppia/app/player/audio/AudioFragmentTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/player/audio/AudioFragmentTest.kt]: Should this toFloat() happen before the division? Otherwise you're losing your fractional component.

Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/app/player/audio/AudioFragmentTest.kt]: Also regarding the name, this is actually a ratio or a percentage not a position. A clearer name is a bit harder to come up with, though.

File: app/src/sharedTest/java/org/oppia/app/player/audio/AudioFragmentTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/player/audio/AudioFragmentTest.kt]: Nit here & below: please add spacing around the binary operators.

File: app/src/sharedTest/java/org/oppia/app/player/audio/AudioFragmentTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/player/audio/AudioFragmentTest.kt]: Nit: please add in-line comments with parameter names for these coordinates.

File: domain/src/main/assets/welcome.json
Comment (BenHenning) [inline domain/src/main/assets/welcome.json]: Here & below: I suggest setting the sizes for when we might use them in the future.

File: domain/src/main/java/org/oppia/domain/audio/AudioPlayerController.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/audio/AudioPlayerController.kt]: Nit: ``} catch (e: IOException) {``

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationRetriever.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/exploration/ExplorationRetriever.kt:20]: Why this change?

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationRetriever.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/exploration/ExplorationRetriever.kt]: Nit: please ensure spacing on each side of binary operators. Ditto elsewhere.

**Review by BenHenning — APPROVED** [review]
Thanks @jamesxu0! Pretty much only nits. Please feel free to submit this once my comments are addressed.

File: domain/src/main/java/org/oppia/domain/exploration/ExplorationRetriever.kt
Comment (jamesxu0) [reply domain/src/main/java/org/oppia/domain/exploration/ExplorationRetriever.kt:20]: I changed this to be the actual corresponding exploration id to the recording filenames, so I was able to test that creating the audio url works.

**Pull request merged**
